### PR TITLE
Gmarchetti/time tracking logic in kurtosis api

### PR DESCRIPTION
**Pull request by galenmarchetti**
goes with https://github.com/kurtosis-tech/kurtosis-core/pull/176

### CODE REVIEW

File: golang/go.mod [hunk 1]
```diff
@@ -4,13 +4,11 @@ go 1.13
...
 require (
 	github.com/golang/protobuf v1.4.2
-	github.com/google/uuid v1.1.2
-	github.com/hashicorp/go-retryablehttp v0.6.7
 	github.com/palantir/stacktrace v0.0.0-20161112013806-78658fd2d177
-	github.com/powerman/rpc-codec v1.2.2
 	github.com/sirupsen/logrus v1.6.0
 	github.com/stretchr/testify v1.6.1
-	golang.org/x/crypto v0.0.0-20201221181555-eec23a3978ad
+	golang.org/x/net v0.0.0-20190404232315-eb5bcb51f2a3 // indirect
```
Comment (galenmarchetti) [inline golang/go.mod:10]: @mieubrisse I just noticed these changes - any idea what happened here?

Comment (mieubrisse) [reply golang/go.mod:10]: Yep, not a cause for concern - basically, we used to use the `crypto` library in Libs, but I pushed the logic inside Core so now it's deleting the deps

File: golang/lib/execution/test_suite_executor.go [hunk 2]
```diff
@@ -102,10 +102,11 @@ func runSerializeSuiteMetadataFlow(ctx context.Context, testsuite testsuite.Test
 		for _, artifactUrl := range testConfig.FilesArtifactUrls {
 			usedArtifactUrls[artifactUrl] = true
 		}
-
 		testMetadata := &core_api_bindings.TestMetadata{
 			IsPartitioningEnabled: testConfig.IsPartitioningEnabled,
 			UsedArtifactUrls:      usedArtifactUrls,
+			TestSetupTimeout: testConfig.TestSetupTimeout,
```
Comment (mieubrisse) [inline golang/lib/execution/test_suite_executor.go:108]: Units

File: golang/lib/execution/test_suite_executor.go [hunk 3]
```diff
@@ -142,12 +143,8 @@ func runTestExecutionFlow(ctx context.Context, testsuite testsuite.TestSuite, co
...
 
 	// Kick off a timer with the API in case there's an infinite loop in the user code that causes the test to hang forever
-	// TODO this should just be "register test execution started", since the API container already has the metadata
-	hardTestTimeout := test.GetExecutionTimeout() + test.GetSetupTeardownBuffer()
-	hardTestTimeoutSeconds := uint64(hardTestTimeout.Seconds())
-	registerTestExecutionMessage := &core_api_bindings.RegisterTestExecutionArgs{TimeoutSeconds: hardTestTimeoutSeconds}
-	if _, err := executionClient.RegisterTestExecution(ctx, registerTestExecutionMessage); err != nil {
-		return stacktrace.Propagate(err, "An error occurred registering the test execution with the API container")
+	if _, err := executionClient.RegisterTestSetup(ctx, &emptypb.Empty{}); err != nil {
+		return stacktrace.Propagate(err, "An error occurred registering the test setup with the API container")
```
Comment (mieubrisse) [inline golang/lib/execution/test_suite_executor.go:147]: Sooo much better

File: golang/lib/execution/test_suite_executor.go [hunk 4]
```diff
@@ -157,41 +154,31 @@ func runTestExecutionFlow(ctx context.Context, testsuite testsuite.TestSuite, co
...
 		filesArtifactUrls)
 
-	// TODO Also time out the setup with the API container rather than storing this locally
-	//  to reduce complexity inside the lib
 	logrus.Info("Setting up the test network...")
 	untypedNetwork, err := test.Setup(networkCtx)
 	if err != nil {
 		return stacktrace.Propagate(err, "An error occurred setting up the test network")
 	}
+	time.Sleep(10 * time.Second)
```
Comment (mieubrisse) [inline golang/lib/execution/test_suite_executor.go:162]: Looks like a debugging holdover

File: golang/lib/execution/test_suite_executor.go [hunk 5]
```diff
@@ -157,41 +154,31 @@ func runTestExecutionFlow(ctx context.Context, testsuite testsuite.TestSuite, co
...
 
 	logrus.Infof("Executing test '%v'...", testName)
+
+	if _, err := executionClient.RegisterTestExecution(ctx, &emptypb.Empty{}); err != nil {
+		return stacktrace.Propagate(err, "An error occurred registering the test execution with the API container")
+	}
 	testResultChan := make(chan error)
 
 	go func() {
 		testResultChan <- runTestInGoroutine(test, untypedNetwork)
```
Comment (mieubrisse) [inline golang/lib/execution/test_suite_executor.go:176]: You can get rid of the multithreading entirely - take the contents of `runTestInGoroutine` and inline them here!

File: golang/lib/execution/test_suite_executor.go [hunk 6]
```diff
@@ -157,41 +154,31 @@ func runTestExecutionFlow(ctx context.Context, testsuite testsuite.TestSuite, co
...
-	select {
-	case testResultErr = <- testResultChan:
-		logrus.Tracef("Test returned result before timeout: %v", testResultErr)
-		timedOut = false
-	case <- time.After(testTimeout):
-		logrus.Tracef("Hit timeout %v before getting a result from the test", testTimeout)
-		timedOut = true
-	}
-	logrus.Tracef("After running test w/timeout: resultErr: %v, timedOut: %v", testResultErr, timedOut)
+	testResultErr := <- testResultChan
```
Comment (mieubrisse) [inline golang/lib/execution/test_suite_executor.go:179]: No more channels needed now!

File: golang/lib/testsuite/test.go [hunk 7]
```diff
@@ -35,18 +34,4 @@ type Test interface {
 	 */
 	Run(network networks.Network, testCtx TestContext)
 
-	/*
```
Comment (mieubrisse) [inline golang/lib/testsuite/test.go:38]: Ooooooo you're doing what I've been wanting to do by moving this into the test config! I like it, though I'd recommend we not do it in this PR, with the reasons being that:
1) it won't be obvious to new users that there are two TestConfiguration values that they _must_ set (the other two, networkPartitioning and filesArtifactUrls, will correctly default to  nothing). To do this properly, I want to have a TestConfigurationBuilder that will default the timeouts to a reasonable value (e.g. 60s) 
2) doing the change will require docs updates

File: golang/lib/execution/test_suite_executor.go (hunk 5 above)
Comment (galenmarchetti) [reply golang/lib/execution/test_suite_executor.go:176]: @mieubrisse if we do that, we miss that nice panic checking that we had in the gofunc (at least, that was my understanding). is that okay?

File: golang/lib/execution/test_suite_executor.go (hunk 6 above)
Comment (galenmarchetti) [reply golang/lib/execution/test_suite_executor.go:178]: same comment as above...my understanding is that we get nice error checking on a panic in test execution with this gofunc technique, but I could be totally off here. lmk if we'd miss out by making this change

File: golang/lib/execution/test_suite_executor.go (hunk 5 above)
Comment (mieubrisse) [reply golang/lib/execution/test_suite_executor.go:175]: Yep so we definitely need to keep the panic guard - the function wouldn't go away; just be renamed to `runTest` - but you can then replace:
```
go func() {
  testResultChan <- runTestInGoroutine(test, untypedNetwork)
  }
  ```
  with simply `runTest(test, untypedNetwork)` + checking the result of the function

**Pull request merged**